idea/idea-fir/src/org/jetbrains/kotlin/idea/intentions/fir/ImportAllMembersIntention.kt

```diff
@@ -0,0 +1,113 @@
+    override val applicabilityRange: HLApplicabilityRange<KtExpression> get() = ApplicabilityRanges.SELF
+
+    override val applicator: HLApplicator<KtExpression, Input> get() = Companion.applicator
+
+    override val inputProvider: HLApplicatorInputProvider<KtExpression, Input> = inputProvider { psi ->
+        val target = psi.reference()?.resolveToSymbol() as? KtNamedClassOrObjectSymbol ?: return@inputProvider null
+        val classId = target.classIdIfNonLocal ?: return@inputProvider null
+        val selectionExpression = (psi.parent as? KtDotQualifiedExpression)?.selectorExpression
+        if (target.origin != KtSymbolOrigin.JAVA &&
+            (target.classKind == KtClassKind.OBJECT ||
+                    // One cannot use on-demand import for properties or functions declared inside objects
+                    isReferenceToObjectMemberOrUnresolved(selectionExpression))
+        ) {
+            // Import all members of an object is not supported by Kotlin.
+            return@inputProvider null
+        }
+        val shortenCommand = collectPossibleReferenceShortenings(
+            psi.containingKtFile,
+            classShortenOption = {
+                val queryClassId = it.classIdIfNonLocal ?: return@collectPossibleReferenceShortenings ShortenOption.DO_NOT_SHORTEN
+                if (queryClassId.packageFqName == classId.packageFqName && queryClassId.relativeClassName.parent() == classId.relativeClassName) {
+                    ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                } else {
+                    ShortenOption.DO_NOT_SHORTEN
+                }
+            },
+            callableShortenOption = {
+                val containingClassId = if (it is KtConstructorSymbol) {
+                    it.containingClassIdIfNonLocal?.outerClassId
+                } else {
+                    it.callableIdIfNonLocal?.classId
+                }
+                if (containingClassId == classId) {
+                    ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                } else {
+                    ShortenOption.DO_NOT_SHORTEN
+                }
+            }
+        )
+        if (shortenCommand.isEmpty) return@inputProvider null
+        Input(classId.asSingleFqName(), shortenCommand)
+    }
+
+    class Input(val fqName: FqName, val shortenCommand: ShortenCommand) : HLApplicatorInput
+
+    companion object {
+        val applicator = applicator<KtExpression, Input> {
+            familyName(KotlinBundle.lazyMessage("import.members.with"))
+            actionName { _, input -> KotlinBundle.message("import.members.from.0", input.fqName.asString()) }
+            isApplicableByPsi { it.isReceiverExpression() && !it.isInImportDirective() }
+            applyTo { _, input ->
+                input.shortenCommand.invokeShortening()
+            }
+        }
+
+        private fun KtExpression.isReceiverExpression(): Boolean {
```

Comment:
Nit: this can be `val`. Also, I wonder if `isExpressionWithReceiver` seems more explicit?

Reply:
`isExpressionWithReceiver` doesn't sound right. This is to check if this expressioin is at the "receiver" expression of its parent. Let me give it a more verbose name.